sympy/solvers/solveset.py

```diff
@@ -998,6 +1003,148 @@ def _solveset(f, symbol, domain, _check=False):
 
     return result
 
+def _is_modular(f, symbol):
+    "Helper function to check solvable modular equations"
+    if not f.has(Mod):
+        return False
+    # extract modterm from f.
+    mt = list(f.atoms(Mod))
+    return len(mt) == 1 and mt[0].has(symbol) and mt[0].args[1].is_Integer
+
+
+def _invert_modular(modterm, rhs, symbol):
+    """
+    Helper function to invert modular equation.
+    ``Mod(a, m) - rhs = 0``
+
+    Generally it is inverted as (a, ImageSet(Lambda(n, m*n + rhs % m), S.Integers)).
+    More simplified form will be returned if possible.
+
+    Parameters
+    ==========
+
+    modterm, rhs : Expr
+        The modular equation to be inverted, ``modterm - rhs = 0``
+
+    symbol : Symbol
+        The variable in the equation to be inverted.
+
+    Returns
+    =======
+
+    A tuple (f_x, y) is being returned where f_x is modular independent function
+    of symbol and y being set of values f_x can have.
+
+    """
+    a, m = modterm.args
+    n = Dummy('n')
+
+    if a is symbol:
+        return symbol, ImageSet(Lambda(n, m*n + rhs % m), S.Integers)
+
+    if a.is_Add:
+        # g + h = a
+        g, h = a.as_independent(symbol)
+        if g is not S.Zero:
+            return _invert_modular(Mod(h, m), (rhs - Mod(g, m)) % m, symbol)
+
+    if a.is_Mul:
+        # g*h = a
+        g, h = a.as_independent(symbol)
+        if g is not S.One:
+            return _invert_modular(Mod(h, m), (rhs*invert(g, m)) % m, symbol)
+
+    if a.is_Pow:
+        # base**expo = a
+        base, expo = a.args
+        if expo.has(symbol) and not base.has(symbol):
+            return expo, FiniteSet(discrete_log(m, rhs, a.base))
+
+    return a, ImageSet(Lambda(n, m*n + rhs % m), S.Integers)
+
+
+def _solve_modular(f, symbol, domain):
+    r"""
+    Helper function for solving modular equations of type ``A - Mod(B, C) = 0``,
+    where A can or cannot be a function of symbol, B is surely a function of
+    symbol and C is an integer.
+
+    Currently only cases when A is not a function of symbol is been implemented.
+
+    Parameters
+    ==========
+
+    f : Expr
+        The modular equation to be solved, ``f = 0``
+
+    symbol : Symbol
+        The variable in the equation to be solved.
+
+    domain : Set
+        A set over which the equation is solved.
+
+    Returns
+    =======
+
+    A set of solutions satisfying the given modular equation.
+    A ``ConditionSet`` if the equation is unsolvable.
```

Comment:
In this case:
```
>>> solve_modulo(Mod(sin(x), 7) - 3, x, S.Reals) # not solvable specifically
    ImageSet(Lambda(_n, ConditionSet(x, Eq(-7*_n + sin(x) - 3, 0), Reals)), Integers)
```
I did this because I was trying to give more information rather than giving no information.
As I mentioned in this [comment](https://github.com/sympy/sympy/pull/16976#discussion_r308014029).

Reply:
`Mod(sin(x), 7) - 3` does not seem to have solutions, since `sin(x)` will lie between `-1` and `1` and when modulo by 7 the result won't be greater than `1`, therefore an `EmptySet` should be returned.

However `Mod(sin(x), 7) - 1` is solvable with solutions as `n*pi  ± pi/2`

Reply:
Currently SymPy is not able to solve the expressions `sin(x) - a*n - b = 0`. What you are saying is theoretically and logically absolutely correct. I added `ConditionSet` now.